src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp

```diff
@@ -65,21 +71,77 @@ MergeTreeReadTaskPtr MergeTreeReadPoolParallelReplicasInOrder::getTask(size_t ta
     std::lock_guard lock(mutex);
 
     if (task_idx >= per_part_infos.size())
-        throw Exception(ErrorCodes::LOGICAL_ERROR,
-            "Requested task with idx {}, but there are only {} parts",
-            task_idx, per_part_infos.size());
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR, "Requested task with idx {}, but there are only {} parts", task_idx, per_part_infos.size());
 
     const auto & part_info = per_part_infos[task_idx]->data_part->info;
-    auto get_from_buffer = [&]() -> std::optional<MarkRanges>
+    const auto & data_settings = per_part_infos[task_idx]->data_part->storage.getSettings();
+    auto get_from_buffer = [&,
+                            rows_granularity = (*data_settings)[MergeTreeSetting::index_granularity],
+                            my_max_block_size = this->block_size_params.max_block_size_rows]() -> std::optional<MarkRanges>
     {
+        const size_t max_marks_in_range = (my_max_block_size + rows_granularity - 1) / rows_granularity;
         for (auto & desc : buffered_tasks)
         {
             if (desc.info == part_info && !desc.ranges.empty())
             {
-                auto result = std::move(desc.ranges);
-                desc.ranges = MarkRanges{};
-                ProfileEvents::increment(ProfileEvents::ParallelReplicasReadMarks, desc.ranges.getNumberOfMarks());
-                return result;
+                if (mode == CoordinationMode::WithOrder)
+                {
+                    /// if already splited, just return desc.ranges
```

Comment:
i think it'd be simpler if we'd maintain `marks_in_range` for each part separately and calculate `split` on the fly. once `marks_in_range[part] >= max_marks_in_range` - we stop subdividing ranges.

Reply:
Great idea, i will change it.